sklearn/tree/_criterion.pyx

```diff
@@ -957,11 +954,107 @@ cdef class MSE(RegressionCriterion):
+        cdef SIZE_t end = self.end
+        compute_weighted_median(dest, start, end, self.sample_weight, self.y,
+                                self.samples, self.y_stride, self.n_outputs)
+        cdef SIZE_t i
+
+    cdef double node_impurity(self) nogil:
+        """Evaluate the impurity of the current node, i.e. the impurity of
+           samples[start:end]"""
+        cdef double* medians = NULL
+        medians = <DOUBLE_t *> calloc(self.n_outputs, sizeof(double))
+        if (medians == NULL):
+            with gil:
+                raise MemoryError()
+        cdef double impurity = 0.0
+        cdef SIZE_t* samples = self.samples
+        cdef SIZE_t i, p, k
+        cdef DOUBLE_t y_ik
+        self.node_value(medians)
+        for k in range(self.n_outputs):
+            for p in range(self.start, self.end):
+                i = samples[p]
+                y_ik = self.y[i * self.y_stride + k]
+                impurity += (fabs(y_ik - medians[k]) / self.weighted_n_node_samples)
+        return impurity / self.n_outputs
+
+    # cdef double proxy_impurity_improvement(self) nogil:
+    #     """Compute a proxy of the impurity reduction
+    #     This method is used to speed up the search for the best split.
+    #     It is a proxy quantity such that the split that maximizes this value
+    #     also maximizes the impurity improvement. It neglects all constant terms
+    #     of the impurity decrease for a given split.
+    #     The absolute impurity improvement is only computed by the
+    #     impurity_improvement method once the best split has been found.
+    #     """
+
+    #     # todo
+    #     pass
+
+    cdef void children_impurity(self, double* impurity_left,
+                                double* impurity_right) nogil:
+        """Evaluate the impurity in children nodes, i.e. the impurity of the
+           left child (samples[start:pos]) and the impurity the right child
+           (samples[pos:end]).
+        """
+        cdef DOUBLE_t* y = self.y
+        cdef DOUBLE_t* sample_weight = self.sample_weight
+        cdef SIZE_t* samples = self.samples
+
+        cdef SIZE_t start = self.start
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
+
+        cdef SIZE_t i, p, k
+        cdef DOUBLE_t y_ik
+
+        cdef double* medians = NULL
+
+        medians = <double*> calloc(self.n_outputs, sizeof(double))
+        if (medians == NULL):
+            with gil:
+                raise MemoryError()
+
+        compute_weighted_median(medians, start, pos, self.sample_weight,
+                                self.y, self.samples, self.y_stride,
+                                self.n_outputs)
+        impurity_left[0] = 0.0
+        impurity_right[0] = 0.0
+
+        for k in range(self.n_outputs):
+            for p in range(start, pos):
+                i = samples[p]
+                y_ik = y[i * self.y_stride + k]
+                # impurity_left[0] += (fabs(y_ik - medians[k]) / (pos - start))
```

Comment:
I'm not sure why this works, but it does; if i do it with the statements that are commented out, there's a float rounding error.